[JSC][ARMv7] ARMv7 WasmMemory::growShared is incorrect

Constellation opened the pull request:
#### 12ed8dc6d02049db08d5bfef98402418f3759c27
<pre>
[JSC][ARMv7] ARMv7 WasmMemory::growShared is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=242600">https://bugs.webkit.org/show_bug.cgi?id=242600</a>
&lt;rdar://96836953&gt;

Reviewed by Mark Lam.

ARMv7 does not implement signaling memory, but it does not fully disable the implementation.
But in this case, we should throw an error instead of crashing since this is exposed function.

* JSTests/wasm/stress/shared-memory-grow-non-crash.js: Added.
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::growShared):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.cpp:
(JSC::JSWebAssemblyMemory::grow):

Canonical link: <a href="https://commits.webkit.org/252395@main">https://commits.webkit.org/252395@main</a>
</pre>

mikhailramalho commented on JSTests/wasm/stress/shared-memory-grow-non-crash.js:
```diff
@@ -0,0 +1,4 @@
+try {
+    new WebAssembly.Memory({initial : 0, maximum : 1, shared : true}).grow(1)
+} catch {
+}
```
The `try-catch` block can be removed. I guess `cvise` was just not smart enough to remove it in the end.

Constellation replied:
@mikhailramalho Since ARM throws an error, we need to keep this try-catch to make every JSC tests pass on all platforms.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmMemory.cpp:
```diff
@@ -494,14 +494,15 @@ bool Memory::addressIsInGrowableOrFastMemory(void* address)
     // in case it's used by multiple threads. Once the signal handler are
     // available, this can be relaxed.
-    RELEASE_ASSERT_NOT_REACHED();
+    return makeUnexpected(GrowFailReason::GrowSharedUnavailable);
```
Would this cause a compiler warning / error due to the rest of this function not being reachable?

Constellation replied:
I don't know since I don't have ARMv7, but so, for now, lets put the rest in #else.

Constellation replied:
Ah, I wonder if this makes code more and more complicated. I changed this `if` to `if 1` and no compiler warning happens. So I'll keep it.

MenloDorian reviewed (approved):
r=me if the compiler is fine with the unreachable code.  Otherwise, probably need to put them the #else case, and also add some UNUSED_PARAMs.

justinmichaud reviewed (commented):
This is disabled by default on shipping ARMv7 browsers, right? LGTM

Constellation committed:
[JSC][ARMv7] ARMv7 WasmMemory::growShared is incorrect
https://bugs.webkit.org/show_bug.cgi?id=242600
<rdar://96836953>

Reviewed by Mark Lam.

ARMv7 does not implement signaling memory, but it does not fully disable the implementation.
But in this case, we should throw an error instead of crashing since this is exposed function.

* JSTests/wasm/stress/shared-memory-grow-non-crash.js: Added.
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::growShared):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.cpp:
(JSC::JSWebAssemblyMemory::grow):

Canonical link: https://commits.webkit.org/252395@main
